benchmarks/dynamo/common.py

```diff
@@ -247,30 +247,40 @@ def tensor_is_on_xla(tensors):
     return any(map(lambda x: x.device.type == "xla", tensors))
 
 
-def timed(model, model_iter_fn, example_inputs, times=1, return_result=False):
+def timed(model, model_iter_fn, example_inputs, times=1, return_result=False, collect_outputs=False):
+    use_xla = tensor_is_on_xla(example_inputs)
     synchronize()
-    if tensor_is_on_xla(example_inputs):
+
+    reset_rng_state()
+    if use_xla:
         import torch_xla.core.xla_model as xm
 
         xm.mark_step()
+        xm.wait_device_ops()
 
-    reset_rng_state()
     t0 = time.perf_counter()
     # Dont collect outputs to correctly measure timing
     for _ in range(times):
-        result = model_iter_fn(model, example_inputs, collect_outputs=False)
-        if tensor_is_on_xla(result):
+        if use_xla:
+            # force a const seed for xla
+            xm.set_rng_state(23, str(xm.xla_device()))
+        result = model_iter_fn(model, example_inputs, collect_outputs=collect_outputs)
+
+        # instead of calling sync on result_list, we should call mark_step.
+        # In training case, result_list may be empty, but we want to sync
+        # all the pending computations.
+        if use_xla:
             # If the model is on XLA device, it's possible that after running
             # the model, the computation is accumulated but not performed yet.
             # Flush all the accumulated computations to make the time measurement
             # accurate.
-            import torch_xla
+            import torch_xla.core.xla_model as xm
+            xm.mark_step()
```

Comment:
i want a better explanation for this mark_step.

1 is to clarify the language a bit- we're only flushing the lazy graph to start being executed, we're not syncing that execution finishes - so this isn't enough to ensure timing is accurate

2 is why we need a mark step here at all; if we're using dynamo backend, we already scheduled a computation that should include most ops.  this mark_step is to capture extra ops for example in aot_autograd epilogue, is that right?

Reply:
i wonder if this mark_step should be part of the backend.  can we install it in the epilogue of aot_autograd after the input accumulation?  cc @bdhirsh i wonder if we can have a way to run some kind of 'epilogue' lambda provided by the backend, after the aot_autograd epilogue finishes.  i also don't think it's great if we have to put xla mark step inside aot_autograd.py

Reply:
also why not use 'maybe_mark_step' here, if it's available?

Reply:
yep. For the `trace_once` bridge, we need `mark_step` for the aot_autograd epilogue and the optimizers; for baseline, we also need this `mark_step`

Reply:
>  For the trace_once bridge, we need mark_step for the aot_autograd epilogue and the optimizers;

Just to confirm: the real reason we need mark_step() is because it's part of the contract with XLA and optimizers, right? When you're using running a program with the xla device and are ready to run `optimizer.step()`, XLA requires you to manually call `mark_step()` first to compile and execute any pending compute.

>  wonder if we can have a way to run some kind of 'epilogue' lambda provided by the backend, after the aot_autograd epilogue finishes

I'm mostly asking the above, because it feels like `mark_step() is something that should live in user code, and not e.g. an aot autograd hook. And our benchmark suite is "technically" user code. But tell me if you agree with that reasoning.

Reply:
`maybe_mark_step` takes parser args which is not available in timed